src/darwin/Framework/CHIP/MTROTAUnsolicitedBDXMessageHandler.mm

```diff
@@ -0,0 +1,142 @@
+/**
+ *
+ *    Copyright (c) 2024 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#import "MTROTAUnsolicitedBDXMessageHandler.h"
+
+#include <protocols/bdx/BdxMessages.h>
+
+using namespace chip;
+using namespace chip::Messaging;
+using namespace chip::bdx;
+
+MTROTAUnsolicitedBDXMessageHandler * MTROTAUnsolicitedBDXMessageHandler::sInstance = nullptr;
+
+CHIP_ERROR MTROTAUnsolicitedBDXMessageHandler::Init(ExchangeManager * exchangeManager)
+{
+    assertChipStackLockedByCurrentThread();
+
+    VerifyOrReturnError(exchangeManager != nullptr, CHIP_ERROR_INCORRECT_STATE);
+
+    MTROTAUnsolicitedBDXMessageHandler::mNumberOfDelegates = 0;
+    mExchangeMgr = exchangeManager;
+    return mExchangeMgr->RegisterUnsolicitedMessageHandlerForProtocol(Protocols::BDX::Id, this);
+}
+
+MTROTAUnsolicitedBDXMessageHandler * MTROTAUnsolicitedBDXMessageHandler::GetInstance()
+{
+    return sInstance;
+}
+
+void MTROTAUnsolicitedBDXMessageHandler::Shutdown()
+{
+    assertChipStackLockedByCurrentThread();
+
+    VerifyOrReturn(mExchangeMgr != nullptr);
+
+    mExchangeMgr->UnregisterUnsolicitedMessageHandlerForProtocol(Protocols::BDX::Id);
+    MTROTAUnsolicitedBDXMessageHandler::mNumberOfDelegates = 0;
+
+    if (mOTAImageTransferHandler != nullptr) {
+        mOTAImageTransferHandler->DestroySelf();
+        delete mOTAImageTransferHandler;
+    }
+}
+
+void MTROTAUnsolicitedBDXMessageHandler::ControllerShuttingDown(MTRDeviceController * controller)
+{
+    assertChipStackLockedByCurrentThread();
+
+    // Since the OTA provider delegate only supports one BDX transfer at a time, calling ShutDown is fine for now.
+    // TODO: #36181 - This class needs to keep a list of all MTROTAImageTransferHandlers mapped to the fabric index and only
+    // delete the MTROTAImageTransferHandler objects with a fabric index matching the MTRDeviceController's fabric index.
+    Shutdown();
```

Comment:
How is this possibly fine?  After this we will stop listening for BDX messages entirely, no?  Was this actually tested with multiple controllers?  Please add to the unit tests to exercise this case.

Reply:
ah the BDX listener should not be shut down. Yeah i need to add tests with multiple controllers

Reply:
fixed this

Reply:
Filed the test part here - https://github.com/project-chip/connectedhomeip/issues/36526. combined in the follow up issue